### DISCUSSION TITLE
Return arrays from stream payloads

### DISCUSSION THREAD

**Discussion by robrichard** [graphql/defer-stream-wg#32] [Architectural Decision Records]
## Context
In the [Feb 2022 WG meeting](https://github.com/graphql/graphql-wg/blob/main/notes/2022/2022-02-03.md#update-on-defer--stream-20m-rob-richard) @leebyron & @yaacovcr proposed changing the stream payload response. 
1. The `data` field on the payload would change to an array containing the streamed list element
2. A new field `atIndex` is added with the array index.
3. The `path` field would point to only the list location, not the index of the list item. 

Initially we would only support one single element returned in the stream array, but we should consider how this payload can be utilized in a future spec amendment to support batches of results or unordered sets.

### Example (containing both defer and stream for completeness):

```graphql
query {
  person(id: "cGVvcGxlOjE=") {
    ...HomeWorldFragment @defer(label: "homeWorldDefer")
    name
    films @stream(initialCount: 2, label: "filmsStream") {
      title
    }
  }
}
fragment HomeWorldFragment on Person {
  homeworld {
    name
  }
}
```

**Response Payloads**

Payload 1
```json
{
  "data": {
    "person": {
      "name": "Luke Skywalker",
      "films": [
        { "title": "A New Hope" },
        { "title": "The Empire Strikes Back" }
      ]
    }
  },
  "hasNext": true
}

```

Payload 2
```json
{
  "label": "homeWorldDefer",
  "path": ["person"],
  "data": {
    "homeworld": {
      "name": "Tatooine"
    }
  },
  "hasNext": true
}
```

Payload 3
```diff
{
  "label": "filmsStream",
- "path": ["person", "films", 2],
+ "path": ["person", "films"],
+ "atIndex": 2,
- "data": {"title": "Return of the Jedi"},
+ "data": [{"title": "Return of the Jedi"}],
  "hasNext": false
}
```

The benefits of this change are:
* `"data": null` always means "error bubbled beyond all fields in this partial-payload" and can never mean "the data for this path was the value null". Valid defer payloads will always have an object set for `data` and now valid stream payloads will always have an array set for `data`. See discussion here: https://github.com/robrichard/defer-stream-wg/discussions/23#discussioncomment-2106541
* This potentially makes it easier to add support for providing batches of results.
* `atIndex` is an additional data point clients can use to differentiate between stream payloads and defer payloads.

## Decision
Discussed at the April 2020 WG meeting and decided to move forward with `data` enclosed in an array in stream payloads. Discussion around how the index is represented will be continued in https://github.com/robrichard/defer-stream-wg/discussions/39

## Consequences
* The data value for a defer payload is an instruction to "merge" at the specified path. Now the data value for a stream payload is an instruction to "concat" at the specified path, instead of providing the value at the specified path.

**Comment by robrichard** [comment]
@josephsavona Any thoughts on this from a Relay perspective?

**Reply by josephsavona** [reply]
> "data": null always means "error bubbled beyond all fields in this partial-payload" and can never mean "the data for this path was the value null".

One interesting data point is that in several years of using the originally proposed behavior at Meta, we have not encountered (to my knowledge) the need to distinguish these cases. There's likely a simple reason, which is that our lists always return non-nullable items (nulls are just pruned). I do wonder if the proposed change here is the right tradeoff: could we instead let developers opt-in to distinguishing these two cases? That said, actually supporting the updated behavior is not _too_ difficult.

> This potentially makes it easier to add support for providing batches of results.

We have implemented support for this at Meta by allowing the server to return an array of GraphQL responses. This approach is more flexible as it allows for defer/stream payloads to be batched together, whereas the proposal here only allows for items from the same `@stream` field to be batched. So I don't think we should consider this as a benefit of this proposal.

> atIndex is an additional data point clients can use to differentiate between stream payloads and defer payloads.

Clients with normalized consistency caches already need to maintain a mapping of outstanding incremental payloads (including the `label`) and then match up new payloads to this data in order to process them (looking up by the response's `label`). If we want to help clients distinguish the type of a payload we should add a `kind` tag: payloads are already highly unstructured (the spec technically allows a payload that is effectively just an `extensions` property, which is useful but also weird). 

Overall this seems fine, it's mostly tedious for Relay because now the OSS spec and our internal implementation differ w/o the difference providing substantial value. However, a better alternative is the named fragments proposal, which allows developers to opt-in to having a named response entity that can be used to check for null-bubbling. I'd much prefer we went with that approach.

**Reply by yaacovCR** [reply]
A point raised at the WG is that if a stream payload can return a batch, that could significantly save on overhead in terms of repeated fields (like path).

Returning an array of multiple payloads on the server, while potentially useful, as far as I can tell does not have this exact benefit. This is a different type of batching for a different purpose, potentially also useful!

i agree that the long term goal should be named fragments or even a normalized response format.

It’s kind of funny that pruning nulls for stream has worked so well! It certainly makes me wonder about the normal nonnull behavior for lists — is there a particular reason why pruning nulls in lists instead of returning errors is still necessary/helpful at Meta when not using stream? I assume that ship has sailed though. It was raised at the WG that the spec could be agnostic about some of these cases and leave it to server implementations to handle as they see best.

**Reply by josephsavona** [reply]
> that could significantly save on overhead in terms of repeated fields (like path).

I'm really skeptical that the amount of bytes saved would be meaningful. What we've seen is that batching for fewer overall flushes (including defer and stream) is useful for mobile clients because it reduces the number of times the antenna has to wake up, saving battery. (i am not an expert in mobile device networking so please consider this a rough analogy). Once you've paid the cost of waking the antenna, a small change in the amount bytes returned doesn't matter as much.

> is there a particular reason why pruning nulls in lists instead of returning errors is still necessary/helpful at Meta when not using stream? 

As a general rule we don't distinguish between different reasons a field may return null: the entity not existing, an internal error, or the viewer having insufficient privileges to view the entity. In some cases we may also include an `errors` entity, but again we are careful to avoid allowing the consumer to distinguish between "does not exist" and "insufficient privileges", since that can be a privacy leak.

With that in mind, most of our list APIs are of the form of connections, where the application shows eg the `first: 3` posts in a user's timeline. If the first 3 posts happen to fail to load (or be inaccessible due to privacy), we don't want to just return `[null, null, null]` to the app, as it would just have to kick off another query to paginate for more. So instead we prune nulls so that we can return the first _non-null_ items that you can actually render. 

> i agree that the long term goal should be named fragments or even a normalized response format.

To this point: if there is consensus on this, i wonder if we can avoid the complexity of the alternate proposal here, and rely on named fragment spreads to provide the null-bubbling boundary.

**Reply by robrichard** [reply]
@josephsavona can you explain a bit more how named fragment spreads could be used to provide null-boundaries for streamed list items?

**Reply by yaacovCR** [reply]
> As a general rule we don't distinguish between different reasons a field may return null: the entity not existing, an internal error, or the viewer having insufficient privileges to view the entity. In some cases we may also include an `errors` entity, but again we are careful to avoid allowing the consumer to distinguish between "does not exist" and "insufficient privileges", since that can be a privacy leak.

Again, really interesting! I believe this behavior is not perfectly in tune with the spec, please-pretty-pleas correct me if I am wrong, as I don't recall or see anything about allowing null swallowing for fields of non-nullable items. Perhaps it should be specifically allowed within the spec to do so. For non-nullable fields that are not enclosed by a list, it would seem that it would be impossibly to speak of "null-swallowing," such that the security concerns you mention seemingly could not be entirely avoided, please correct me if I am wrong.

In any case, seeing as this behavior at Meta is not limited to stream, it would seem that it would deserve a separate discussion, and the stream implementation would have to handle non-swallowed nulls.

**Reply by nodkz** [reply]
I think that we want too much from `stream`. We consider it like `AsyncIterator` with batching - something like `BatchAsyncInterator`. I wrote my thought here about this https://github.com/robrichard/defer-stream-wg/discussions/23#discussioncomment-2155981

**Reply by josephsavona** [reply]
> can you explain a bit more how named fragment spreads could be used to provide null-boundaries for streamed list items?

@robrichard sure. With names fragments at any point in a query you can create a new level of nesting in the response, which provides a point that can be nulled-out independently of its parent:

With the standard approach:
```
foo @stream {
  bar
}
```
If bar is a non-null field and errors, it has to bubble up - leading to the use-case here. But you can just wrap all the fields of the stream to avoid this:

```
foo @stream {
  myNullBoundary: … { 
    bar
  }
}
```

now if bar errors, myNullBoundary gets errored out in the response. This is a clean way to allow users to opt-in to distinguishing “item value was null” and “item was nulled out” if it actually matters to them.

**Reply by josephsavona** [reply]
>  I believe this behavior is not perfectly in tune with the spec, please-pretty-pleas correct me if I am wrong, as I don't recall or see anything about allowing null swallowing for fields of non-nullable items.

@yaacovCR Im fairly sure that this happens in our data layer, not as a default behavior of our GraphQL executor. In any case it’s easy to imagine that this _could_ be implemented a layer below, so that the executor only sees non-null items.

**Reply by yaacovCR** [reply]
A well performing data layer makes a lot more sense than a non compliant executor —- but as graphql aims to provide type/null safety for non compliant data layers, this also suggests we cannot/should not assume nulls will be swallowed…

**Reply by josephsavona** [reply]
@yaacovCR no we shouldn’t assume all nulls will be swallowed, but optimizing for common cases does make sense. See my comment about how named fragment spreads achieve a compromise.

**Reply by yaacovCR** [reply]
I see the compromise but in practice I am still confused. We are trying to make it clear from response only without knowledge of schema whether a list item was null or whether a null bubbled up to the list because the list items are non nullable.

a client might want to know this because null list items would get integrated into the data store, while if the null bubbled up to the list itself, it wouldn’t. A client could check the schema to figure out which one happened, but everyone is seemingly agreeing that we want it to be obvious from the response.

the problem I have is that @josephsavona  suggestion seems to imply that if a client would want to distinguish these cases, it would have to transform all of its streamed fields to use named fragments. Meaning, even though perhaps only some clients would want to distinguish these cases, those clients would presumably want to do so for all operations with streams.

End users creating queries/operations should not need to know anything about how their clients work, and so the burden of adding the error boundary with the named fragment should definitely be on the client and not the query author.

but some clients do not regularly do any query rewriting or manipulation. It would probably be easier just to check the schema and see if the field is nullable than to transform all queries. So this optional method, while theoretically possible seems like so great a burden it would rarely be used.

Perhaps it’s not necessary to distinguish between these cases, null list items and null bubbling up the list purely from within the stream payload response. But if it is worth it, named fragments does not seem like a path that would likely be used.

**Reply by josephsavona** [reply]
> We are trying to make it clear from response only without knowledge of schema

Ah, yeah i think this is the disconnect. The power of GraphQL fundamentally stems from the use of a schema: clients that don't make use of the schema are limited in the capabilities they can provide, and the solutions we come up with in the spec to support them increase redundancy for everyone. I do think it should be _possible_ for schema-ignoring clients to distinguish "value was null" from "value was nulled out" _if that matters_ — but my pushback is whether that should be the default behavior. Named fragments would allow clients that actually care to distinguish.

**Reply by yaacovCR** [reply]
Fwiw the [graphql-executor pr](https://github.com/yaacovCR/graphql-executor/pull/154) now supports both @josephsavona style batching of all available payloads as well as delaying stream payloads until an interval has passed or a certain chunk size has been reached.

It also supports parallel streaming, in which case an atIndices array describes the paths/indices of the corresponding returned results within the data array, instead of just a single atIndex value.

@robrichard @nodkz @josephsavona

**Comment by yaacovCR** [comment]
I think that field resolvers that return an async iterable should return an async iterable of iterables, ie an async iterable that returns chunks of the larger list.

**Reply by yaacovCR** [reply]
The wip graphql executor implementation takes that approach.

https://github.com/yaacovCR/graphql-executor/pull/154

**Reply by IvanGoncharov** [reply]
Makes sense in the context of what was discussed at the last WG.
However, DX should be intuitive and `async iterable of iterables` is not, especially for fields that return multi-dimensional arrays.
Also, we need to support the simplest case of non-batching resolver as simple `async iterable of items`.

Not sure what is a good solution, but here is an idea from the top of my head.
Some class to represent a slice of the array, JS doesn't have one so we can have custom one and support `AsyncIterable<ItemT> | AsyncIterable<GraphQLListValueSlice<ItemT>>` as a return type from resolver.
```
class GraphQLListValueSlice<ItemT> {
  items: Iterable<ItemT>
  
  constructor(items) {
    this.items = items;
  }
}
```
In the future, we can also add `offset` property to support "sparse-resolved async iterables" if stream payload will support "sparse-stream" as was discussed during last WG.

**Reply by yaacovCR** [reply]
I don’t think we need to support simplest case of non batching resolver, we can force implementers to return a batch of size one. Luckily we have not merged the above async iterable code yet, so that is still possible.

My idea for sparse streams that we would have the iterable return an object instead of an array with number keys corresponding to the indices. I am not sure how your class would work with both batching and sparseness, see below.

I don’t think the types are so confusing, even for nested lists, basically the AsyncIterable would (a) return the same type as the field, corresponding to a subset of the list or (b) for sparse streams, would return an object with numbered indices where the values are the same type as the list members.

I don’t think we should allow non batching async iterables

if we do allow them, I agree we would need a class, except I think we would have to modify your suggestion to allow sparseness even  within/across chunks, ie the first chunk might have the first 100 elements except 10 or so that got delayed and are in the next chunk

either way, very important to decide this prior to merging current implementation, considering that I am proposing to require fields to always return an iterable (or an object with numeric keys)

if we reject my suggestion, ok, but thanks for pitching in to discuss as there could be breaking changes if we make the wrong choice

**Reply by yaacovCR** [reply]
@robrichard  any thoughts

**Reply by yaacovCR** [reply]
A class would require instanceof dynamic checks which we currently don’t really have, although we check for returned errors with instanceof, implementers can always throw errors without instanceof check

graphql executor works across realms, we want to use duck typing rather than nominal (if I am using terms correctly)

also, to get batching to work in resolvers, with a class, resolvers would depend on graphql js, not the hugest deal but previously resolvers could be dependency free functions

**Reply by yaacovCR** [reply]
Basically the allowed return types of the async iterable in my vision would correspond to the possible types of the data field in the stream payload when the field is streamed. If it’s not streamed, the executor would roll everything into one list.

**Reply by n1ru4l** [reply]
To me, the solution proposed by @IvanGoncharov in https://github.com/robrichard/defer-stream-wg/discussions/32#discussioncomment-2124057 adds unnecessary complexity that is not really needed.

Having the same well-documented API (always yield an array) feels like a simpler approach than having an overloaded API (either single item or `GraphQLListValueSlice(items)`).

Furthermore, as TypeScript support within this library becomes better the rules could be enforced by stricter type-interference/checks.

**Reply by yaacovCR** [reply]
update: the graphql-executor WIP PR https://github.com/yaacovCR/graphql-executor/pull/154 now supports arbitrary chunkSizes based on the above approach.

**Comment by nodkz** [comment]
I consider `@stream` like `AsyncIterator`. If we want to provide `BatchAsyncInterator` why not create a new directive `@streamBatch(batchSize: 100, batchTimeoutMs: 50, ...rest_stream_args)`.

Let's keep the `data` field in `@stream` with any type (scalar, object, array), but for `@streamBatch` it will be an array of any type.

PS. A little bit more thoughts about `defer == Promise` & `stream == AsyncIterator`: https://github.com/robrichard/defer-stream-wg/discussions/23#discussioncomment-2155981

**Reply by yaacovCR** [reply]
@nodkz see the first and third benefits outlined by @robrichard above at https://github.com/robrichard/defer-stream-wg/discussions/32#discussion-3855329 How do these seem to you, do you have in mind any alternatives?

**Reply by nodkz** [reply]
> "data": null always means "error bubbled beyond all fields in this partial-payload" and can never mean "the data for this path was the value null". Valid defer payloads will always have an object set for data and now valid stream payloads will always have an array set for data. See discussion here: https://github.com/robrichard/defer-stream-wg/discussions/23#discussioncomment-2106541

I consider stream like AsyncIterator. So how does it return an error? Just throw an Error without a `data` field. 

> `atIndex` is an additional data point clients can use to differentiate between stream payloads and defer payloads.

This is a good param. But in terms of AsyncIterator better to name it like `idx` or `seqId`.

**Reply by nodkz** [reply]
If a client wants to work with batching it might be
- streamBatch(maxBatchSize: 100, maxBatchTimeout: 50) - additional directive with specific logic
- or regular `stream` which works like AsyncIterator where batching logic is moved to the schema implementation side, something like:

```graphql
query {
  filmStream @stream() {
      batch(maxSize: 100) {
        title
      }
  }
}
```

Payload 1 (Initial payload)
```json
{
  "data": {
     "filmStream": []
  },
  "hasNext": true
}
```

Payload 2 (first 100 films) 
```json
{
  "path": ["filmStream"],
  "idx": 1,
  "data": {
      "batch": ["Film 1", "Film 2", ..., "Film 100"],
  },
  "hasNext": true
}
```

Payload 3 (second 100 films)
```json
{
  "path": ["filmStream"],
  "idx": 2,
  "data": {
      "batch": ["Film 101", "Film 102", ..., "Film 200"],
  },
  "hasNext": true
}
```

Payload 4 (just 30 records)
```json
{
  "path": ["filmStream"],
  "idx": 3,
  "data": {
      "batch": ["Film 201", "Film 202", ..., "Film 231"],
  },
  "hasNext": false
}
```

Payload 5 (Film 232 produces an error) 🛑
```json
{
  "path": ["filmStream"],
  "idx": 4,
  "errors": [{...}],
  "hasNext": true
}
```

Payload 6 (continue from Film 233 + further 99 records)
```json
{
  "path": ["filmStream"],
  "idx": 5,
  "data": {
      "batch": ["Film 233", "Film 234", ..., "Film 332"],
  },
  "hasNext": false
}
```

**Reply by yaacovCR** [reply]
> I consider stream like AsyncIterator. So how does it return an error? Just throw an Error without a `data` field.

This was the initial suggestion in the working group. The hesitation stemmed -- I think -- from the fact that errors without data might mean different things depending on the context:

for an initial response = errors were encountered prior to execution
for defer payloads = would never happen
for stream payloads = would be a null bubbling up from the list item to the field.

Under the above proposal to return arrays, stream would be similar to defer and there would never be errors without data -- at least as I understand it.

Having said that, you may think that it would be better to just do overload the errors-only case rather than introduce the array, I could see how that in some respects might be simpler, but I tend to think that using a null versus an array is more straightforward.

**Comment by benjie** [comment]
I'm skeptical of the need for `atIndex`.

- Currently we require the payloads to be in order, so `atIndex` is redundant information that inflates the payload size
- Clients need to figure out what to do if the `atIndex` is a value different to that which they expect (i.e. not the length of the array they already have)
- We need to consider what would happen if the server gives an `atIndex` that's massive (e.g. `999999999`) - does the client have to allocate a list that big just to set those entries? Seems like a possible security issue for a client to a malicious GraphQL server

One thing in its favour is it allows for us to do out-of-order delivery later, but I wonder if that's sufficiently likely and sufficiently valuable to justify these tradeoffs.

**Reply by yaacovCR** [reply]
> I'm skeptical of the need for `atIndex`.

I believe @robrichard is in your corner, it can basically be appended t the end of the path

> * Currently we require the payloads to be in order, so `atIndex` is redundant information that inflates the payload size
> [sic] One thing in its favour is it allows for us to do out-of-order delivery later, but I wonder if that's sufficiently likely and sufficiently valuable to justify these tradeoffs.

graphql-executor [already supports](https://github.com/yaacovCR/graphql-executor/pull/154#issuecomment-1083513646) batched and parallel ("out-of order") delivery -- I don't believe the initial release at graphql-js will support this, but it's not too difficult to foresee a future where this is useful and available.

> * Clients need to figure out what to do if the `atIndex` is a value different to that which they expect (i.e. not the length of the array they already have)

Wouldn't that still be true if we put the index back in the path? The atIndex value is the same value that used to be the last entry in the path key.

> * We need to consider what would happen if the server gives an `atIndex` that's massive (e.g. `999999999`) - does the client have to allocate a list that big just to set those entries?

That would seem to be a problem just in general, even if the value currently proposed to be in atIndex was put back into the path.

> Seems like a possible security issue for a client to a malicious GraphQL server

Do you mean *from* a malicious GraphQL server?

**Reply by robrichard** [reply]
Standalone index discussion opened: https://github.com/robrichard/defer-stream-wg/discussions/39

**Reply by benjie** [reply]
> Wouldn't that still be true if we put the index back in the path? The atIndex value is the same value that used to be the last entry in the path key.
> That would seem to be a problem just in general, even if the value currently proposed to be in atIndex was put back into the path.

I don't believe it belongs in the path NOR atIndex. I think it's completely redundant and should be removed. But if it is kept, I would much rather see it as `atIndex` and _not_ in `path`. It definitely feels wrong being in `path` to me.

> > Seems like a possible security issue for a client to a malicious GraphQL server

> Do you mean from a malicious GraphQL server?

I meant:
`Seems like a possible security issue for (a client to a malicious GraphQL server)`

I think maybe you read:
`(Seems like a possible security issue for a client) to a malicious GraphQL server`

English is fun :laughing:

**Reply by benjie** [reply]
I've gone into more detail in my reasoning in the issue Rob linked above: https://github.com/robrichard/defer-stream-wg/discussions/39#discussioncomment-2543343

**Comment by robrichard** [comment]
Discussed at the April 2020 WG meeting and decided to move forward with data enclosed in an array in stream payloads. Discussion around how the index is represented will be continued in https://github.com/robrichard/defer-stream-wg/discussions/39

**Reply by robrichard** [reply]
The defer-stream branch on `graphql-js` is now updated to the behavior we agreed on in the WG meetings:
Stream payloads are now wrapped in `[]` and non-null errors will bubble up to `"data": null`

**Comment by glasser** [comment]
Note for posterity: while the decision to send the data in an array still holds, we are considering renaming the field from `data` to `items` in https://github.com/robrichard/defer-stream-wg/discussions/42
